### PR TITLE
feat: Refine triggering abstraction

### PR DESCRIPTION
**Pull request by robinholzi**
# Motivation

Let's refine the abstraction for trigger a bit to enable more powerful triggering configuration.

# Changes

Note: we will adjust the business logic in the future to use the new abstraction.

- New configuration models & refinement of existing models
- move trigger configs in dedicated files

### CODE REVIEW

File: modyn/config/schema/pipeline/trigger/ensemble.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+
```
Comment (robinholzi) [inline modyn/config/schema/pipeline/trigger/ensemble.py:1]: New, but not yet reflected in the business logic.

File: docs/pipeline/TRIGGERING.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# Modyn Triggering
+
+Alongside the simple triggers, Modyn also provides complex triggers that can be used to trigger the training of a model. 
+
+Despite the line being blurry, complex triggers are generally more sophisticated and require more information to be provided to the trigger - they most of the time cannot be pre-determined via a simple configuration entry but require
+reinjested information from the ongoing pipeline run.
```
Comment (MaxiBoether) [inline docs/pipeline/TRIGGERING.md:6]: reingested :D?

File: docs/pipeline/TRIGGERING.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+Some complex triggers can only make decisions in batched intervals as they are cannot be efficiently computed
+on a per-sample basis. Here we can find the `DataDrift` and `CostBased` (not yet implemented) based triggers.
+
+Another policy type is the `EnsemblePolicy` which can be used to combine multiple triggers into a single trigger. This can be useful if multiple triggers should be evaluated before the training of a model is triggered.
```
Comment (MaxiBoether) [inline docs/pipeline/TRIGGERING.md:11]: Oh ok sure, sounds good, I mean nice to have but no urgent need for this right now I suppose

File: modyn/config/schema/pipeline/config.py [hunk 4]
```diff
@@ -30,7 +30,7 @@ class ModynPipelineConfig(ModynBaseModel):
 
     training: TrainingConfig
     data: DataConfig
-    trigger: TriggerConfig
+    trigger: TriggerConfig  # type: ignore[valid-type]
```
Comment (MaxiBoether) [inline modyn/config/schema/pipeline/config.py:33]: why this change in mypy? it was not complaining, was it?

File: modyn/config/schema/pipeline/trigger/drift.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+__TriggerConfig = ForwardRef("TriggerConfig", is_class=True)
+
+
+class DataDriftTriggerConfig(ModynBaseModel):
+    id: Literal["DataDriftTrigger"] = Field("DataDriftTrigger")
+
+    detection_interval: Optional[__TriggerConfig] = Field(  # type: ignore[valid-type]
+        None, description="The Trigger policy to determine the interval at which drift detection is performed."
+    )  # currently not used
```
Comment (MaxiBoether) [inline modyn/config/schema/pipeline/trigger/drift.py:16]: I don't get this yet. How do you want to pass a trigger_config to this (and why)? can i pass a drift trigger to a drift trigger? could you elaborate a bit here?

File: modyn/config/schema/pipeline/trigger/ensemble.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        Function that aggregates the decisions of the individual triggers."""
+        raise NotImplementedError
+
+
+class MajorityVoteEnsembleStrategy(BaseEnsembleStrategy):
+    id: Literal["MajorityVote"] = Field("MajorityVote")
+
+    @property
+    def aggregate_decision_func(self) -> Callable[[dict[str, bool]], bool]:
+        return lambda decisions: sum(decisions.values()) > len(decisions) / 2
```
Comment (MaxiBoether) [inline modyn/config/schema/pipeline/trigger/ensemble.py:23]: i am somewhat torn here since we now have business logic in the config files, but this is simple logic so it might be ok...

**Review by MaxiBoether — APPROVED** [review]
generally LGTM, had a short question on the new field on the data drift trigger

File: modyn/config/schema/pipeline/config.py (hunk 4 above)
Comment (robinholzi) [reply modyn/config/schema/pipeline/config.py:33]: It was complaining in an earlier version. I think we can remove it now though! Thanks!

File: docs/pipeline/TRIGGERING.md (hunk 2 above)
Comment (robinholzi) [reply docs/pipeline/TRIGGERING.md:6]: I wonder why the typo pre-commit hook didn't detect this😅

File: docs/pipeline/TRIGGERING.md (hunk 3 above)
Comment (robinholzi) [reply docs/pipeline/TRIGGERING.md:11]: Yeah, I probably won't implement this in the next month.

File: modyn/config/schema/pipeline/config.py (hunk 4 above)
Comment (robinholzi) [reply modyn/config/schema/pipeline/config.py:33]: Still required, unfortunately. 
image

File: modyn/config/schema/pipeline/trigger/drift.py (hunk 5 above)
Comment (robinholzi) [reply modyn/config/schema/pipeline/trigger/drift.py:16]: I noticed that drift the trigger only supports to be evaluated after every n-th sample. However, there's no real reason to not use a time schedule for assessing the drift trigger. Consequently, I wanted to support specifying a TimeTrigger or DataAmountTrigger. However, there's no real reason not to support other trigger types to as they follow the same interface. In practise you would want to use a cheaper policy though to determine the evaluation of the more expensive ones.

File: modyn/config/schema/pipeline/trigger/ensemble.py (hunk 6 above)
Comment (robinholzi) [reply modyn/config/schema/pipeline/trigger/ensemble.py:23]: Also considered that. I concluded though that this is the most accessible setup.
